ydb/library/yql/udfs/common/knn/knn.cpp

```diff
@@ -0,0 +1,247 @@
+    };
+    
+    auto enumerateOneSized = [&callback] (const TUnboxedValuePod vector1, const TUnboxedValue* elements1, const TUnboxedValuePod vector2) {
+        const auto size = vector1.GetListLength();
+        ui32 idx = 0;
+        TUnboxedValue value;
+        const auto it = vector2.GetListIterator();          
+
+        while (it.Next(value)) {
+            callback(elements1[idx++].Get<float>(), value.Get<float>());
+        }
+
+        // Length mismatch
+        if (it.Next(value) || idx != size)
+            return false;
+
+        return true;
+    };
+
+    auto enumerateNoSized = [&callback] (const TUnboxedValuePod vector1, const TUnboxedValuePod vector2) {
+        TUnboxedValue value1, value2;
+        const auto it1 = vector1.GetListIterator();
+        const auto it2 = vector2.GetListIterator();    
+        for (; it1.Next(value1) && it2.Next(value2);) {
+            callback(value1.Get<float>(), value2.Get<float>());
+        }
+
+        // Length mismatch
+        if (it1.Next(value1) || it2.Next(value2))
+            return false;
+
+        return true;
+    };
+
+    const auto elements1 = vector1.GetElements();
+    const auto elements2 = vector2.GetElements();
+    if (elements1 && elements2) {
+        if (!enumerateBothSized(vector1, elements1, vector2, elements2))
+            return false;
+    } else if (elements1) {
+        if (!enumerateOneSized(vector1, elements1, vector2))
+            return false;
+    } else if (elements2) {
+        if (!enumerateOneSized(vector2, elements2, vector1))
+            return false;
+    } else {
+        if (!enumerateNoSized(vector1, vector2))
+            return false;
+    }
+
+    return true;
+}
+
+enum EFormat : unsigned char {
+    FloatVector = 1
+};
+
+TString SerializeFloatVector(const TUnboxedValuePod x) {
+    const EFormat format = EFormat::FloatVector;
+    
+    TString str;
+    TStringOutput outStr(str);
+
+    if (const auto elements = x.GetElements()) {
+        const auto size = x.GetListLength();
+        outStr.Reserve(1 + size * sizeof(float));
+    }
+    outStr.Write(&format, sizeof(unsigned char));
+
+    EnumerateVector(x,  [&outStr] (float element) { outStr.Write(&element, sizeof(float)); });
+
+    return str;
+}
+
+NYql::NUdf::TUnboxedValue DeserializeFloatVector(const IValueBuilder *valueBuilder, TStringRef str) {
+    if (str.Size() % sizeof(float) != 0)    
+        return {};
+    
+    const ui32 count = str.Size() / sizeof(float);
+
+    TUnboxedValue* items = nullptr;
+    auto res = valueBuilder->NewArray(count, items);
+    
+    TMemoryInput inStr(str);
+    for (ui32 i = 0; i < count; ++i) {
+        float element;
+        if (inStr.Read(&element, sizeof(float)) != sizeof(float))
+            return {};
+        *items++ = TUnboxedValuePod{element};
+    }
+
+    return res.Release();
+}
+
+SIMPLE_STRICT_UDF(TToBinaryString, char*(TAutoMap<TListType<float>>)) {
+    return valueBuilder->NewString(SerializeFloatVector(args[0]));
```

Comment:
please avoid intermediate TString as result of SerializeFloatVector, pass valueBuilder directly